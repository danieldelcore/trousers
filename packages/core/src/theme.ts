// @ts-ignore
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function createTheme(theme: Record<string, any>) {
    return '';
}

export default createTheme;
