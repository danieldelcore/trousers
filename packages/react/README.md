# @trousers/react

`@trousers/react` is a Trousers implementation specifically for React.

## Installation

`yarn add @trousers/react` or `npm install @trousers/react`

## API

### `jsx`

This package exports a custom `jsx` pragma which can be used to enable the CSS prop.

-   The pragma allows for HTML elements to accept a `css` prop and modifier props, denoted by the `$` symbol.
-   The pragma contains logic for mounting and processing the styles for use with React
-   Modifiers accept a boolean value. Under the hood will toggle the relevant styles on or off accordingly. (if supplied by the styles passed into the css prop).

**Example: Basic usage**

```jsx
/** @jsx jsx */
import css from '@trousers/core';
import jsx from '@trousers/react';

const styles = css('button', { backgroundColor: 'red' }).modifier('primary', {
    backgroundColor: 'blue',
});

const Button = ({ primary }) => <button css={styles} $primary={primary} />;
```
