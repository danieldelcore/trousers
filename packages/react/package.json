{
    "name": "@trousers/react",
    "version": "4.0.0",
    "main": "dist/trousers-react.cjs.js",
    "module": "dist/trousers-react.esm.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {
        "@trousers/core": "3.0.2",
        "@trousers/sheet": "3.0.0"
    },
    "devDependencies": {
        "react": "^17.0.1"
    },
    "peerDependencies": {
        "react": ">=16.8.0"
    },
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/react"
}
