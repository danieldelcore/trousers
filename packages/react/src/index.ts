// eslint-disable-next-line prettier/prettier
export type { TrousersProps } from './types';

export { default } from './jsx';
