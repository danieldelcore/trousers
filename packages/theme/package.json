{
    "name": "@trousers/theme",
    "version": "3.0.0",
    "main": "dist/index.js",
    "module": "dist/index.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {
        "@trousers/hash": "3.0.0"
    },
    "devDependencies": {
        "react": "^16.8.1"
    },
    "peerDependencies": {
        "react": ">=16.8.0"
    },
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/theme"
}
