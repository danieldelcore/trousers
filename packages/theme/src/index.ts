export {
    ThemeProvider,
    ThemeConsumer,
    ThemeContext,
    ThemeCtx,
    ThemeProviderProps,
} from './ThemeContext';
export { default as useTheme } from './useTheme';
