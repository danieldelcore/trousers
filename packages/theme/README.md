# @trousers/theme

`@trousers/theme` provides tools for theming your application.

Theming is achieved via React's Context API. Use the `ThemeProvider` to send your theme down the react tree and give all usages of `useStyles` and `useGlobals` access to your theme. You can even choose to nest themes and present a section of your app in a different way.

It looks a little something like this:

```jsx
import { ThemeProvider } from 'trousers';

const lightTheme = {
    primaryColor: 'white',
    secondaryColor: 'blue',
    disabledColor: 'grey',
};

const darkTheme = {
    primaryColor: 'black',
    secondaryColor: 'purple',
    disabledColor: 'grey',
};

const MyApp = () => {
    return (
        <ThemeProvider theme={lightTheme}>
            <h1>Hello World</h1>
            <p>Rest of my app lives here and has access to the light theme!</p>
            <ThemeProvider theme={darkTheme}>
                <p>This subtree will have access to the dark theme!</p>
            </ThemeProvider>
        </ThemeProvider>
    );
};
```

When a Trousers component is mounted within a new theme context, it will render new styles and apply them to the component.

You can define how your component handles themes like this:

```jsx
const buttonStyles = props => styleCollector('button').element`
        background-color: ${theme => theme.secondaryColor};
    `.modifier(props.primary)`
        background-color: ${theme => theme.primaryColor};
    `.modifier(props.disabled)`
        background-color: ${theme => theme.disabledColor};
    `;
```

Now your component will render different styles based on the context it is mounted in.

## API

### `<ThemeProvider />`

Responsible for pushing the supplied theme into React's Context API.

**Props:**

-   `theme`: Object

**Example:**

```jsx
import React from 'react';
import { ThemeProvider } from '@trousers/theme';

const theme = {
    primaryColor: 'red',
    secondaryColor: 'blue',
};

const App = () => (
    <ThemeProvider theme={theme}>
        {* Every child node will have access to the theme *}
    </ThemeProvider>
);
```

### `useTheme`

A handy utility hook which pulls the theme from context. Useful when you want to use the CSS prop with object notation.

```jsx
/** @jsx jsx */
import React from 'react';
import { jsx } from '@trousers/core';
import { ThemeProvider, useTheme } from '@trousers/theme';

const theme = {
    primaryColor: 'red',
    secondaryColor: 'blue',
};

const Foo = () => {
  const theme = useTheme();

  return (
    <div css={{
      color: theme.primaryColor;
    }}>
      Foo bar
    </div>
  );
}

const App = () => (
    <ThemeProvider theme={theme}>
      <Foo />
    </ThemeProvider>
);
```
