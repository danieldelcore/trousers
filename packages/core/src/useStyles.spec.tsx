import React, { FC, Fragment } from 'react';
import { render, cleanup } from '@testing-library/react';
import { ThemeProvider } from '@trousers/theme';
import styleCollector from '@trousers/collector';

import { css, useStyles } from './';

interface Props {
    isRed?: boolean;
    isBlue?: boolean;
}

describe('useStyles', () => {
    afterEach(() => {
        document.getElementsByTagName('html')[0].innerHTML = '';

        cleanup();
    });

    it('returns correct className', () => {
        const styles = styleCollector('foo').element`
            background-color: blue;
        `;

        const StyledComponent = () => {
            const classNames = useStyles(styles);

            return <span className={classNames}>Ahoy!</span>;
        };

        const { container } = render(<StyledComponent />);

        expect(container.querySelector('span')!.className).toEqual(
            'foo__38890833250',
        );
    });

    it('returns correct classNames for multiple modifiers', () => {
        const styles = (props: Props) => styleCollector('foo').element`
            background-color: green;
        `.modifier(props.isRed)`
            background-color: red;
        `.modifier(props.isBlue)`
            background-color: blue;
        `;

        const StyledComponent: React.FC<Props> = props => {
            const classNames = useStyles(styles(props));

            return <span className={classNames}>Ahoy!</span>;
        };

        {
            const { container } = render(<StyledComponent />);

            expect(container.querySelector('span')!.className).toEqual(
                'foo__38262462060',
            );
        }
        {
            const { container } = render(<StyledComponent isRed />);

            expect(container.querySelector('span')!.className).toEqual(
                'foo__38262462060 foo--4911895800',
            );
        }
        {
            const { container } = render(<StyledComponent isBlue />);

            expect(container.querySelector('span')!.className).toEqual(
                'foo__38262462060 foo--38890833250',
            );
        }
        {
            const { container } = render(<StyledComponent isBlue isRed />);

            expect(container.querySelector('span')!.className).toEqual(
                'foo__38262462060 foo--4911895800 foo--38890833250',
            );
        }
    });

    it('returns correct classNames for multiple named modifiers', () => {
        const styles = (props: Props) => styleCollector('foo').element`
            background-color: blue;
        `.modifier('primary', props.isRed)`
            background-color: red;
        `.modifier('secondary', props.isBlue)`
            background-color: isBlue;
        `;

        const StyledComponent: React.FC<Props> = props => {
            const classNames = useStyles(styles(props));

            return <span className={classNames}>Ahoy!</span>;
        };

        const { container } = render(<StyledComponent />);

        expect(container.querySelector('span')!.className).toEqual(
            'foo__38890833250',
        );

        const { container: secondContainer } = render(
            <StyledComponent isRed />,
        );

        expect(secondContainer.querySelector('span')!.className).toEqual(
            'foo__38890833250 foo--primary4911895800',
        );

        const { container: thirdContainer } = render(
            <StyledComponent isBlue />,
        );

        expect(thirdContainer.querySelector('span')!.className).toEqual(
            'foo__38890833250 foo--secondary1129918790',
        );

        const { container: forthContainer } = render(
            <StyledComponent isRed isBlue />,
        );

        expect(forthContainer.querySelector('span')!.className).toEqual(
            'foo__38890833250 foo--primary4911895800 foo--secondary1129918790',
        );
    });

    it('attaches styles to the head', () => {
        const styles = styleCollector('foo').element`
            background-color: blue;
        `;

        const StyledComponent = () => {
            const classNames = useStyles(styles);

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches styles with modifiers to the head', () => {
        const styles = (props: Props) => styleCollector('foo').element`
            background-color: blue;
        `.modifier(props.isRed)`
            background-color: red;
        `;

        const StyledComponent: React.FC<Props> = props => {
            const classNames = useStyles(styles(props));

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(<StyledComponent isRed />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches styles with named modifiers to the head', () => {
        const styles = (props: Props) => styleCollector('foo').element`
            background-color: blue;
        `.modifier('primary', props.isRed)`
            background-color: red;
        `;

        const StyledComponent: React.FC<Props> = props => {
            const classNames = useStyles(styles(props));

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(<StyledComponent isRed />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches styles with multiple named modifiers to the head', () => {
        const styles = (props: Props) => styleCollector('foo').element`
            background-color: blue;
        `.modifier('primary', props.isRed)`
            background-color: red;
        `.modifier('secondary', props.isBlue)`
            background-color: isBlue;
        `;

        const StyledComponent: React.FC<Props> = props => {
            const classNames = useStyles(styles(props));

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(<StyledComponent isRed />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(<StyledComponent isBlue />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(<StyledComponent isRed isBlue />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches themed styles to the head', () => {
        interface Theme {
            backgroundColor: string;
            backgroundColorSecondary: string;
        }

        const theme: Theme = {
            backgroundColor: 'purple',
            backgroundColorSecondary: 'green',
        };

        const styles = (props: Props) => styleCollector<Theme>('foo').element`
            background-color: ${theme => theme.backgroundColor};
        `.modifier(props.isRed)`
            background-color: ${theme => theme.backgroundColor};
        `;

        const StyledComponent: React.FC<Props> = props => {
            const classNames = useStyles(styles(props));

            return <span className={classNames}>Ahoy!</span>;
        };

        render(
            <ThemeProvider theme={theme}>
                <StyledComponent />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(
            <ThemeProvider theme={theme}>
                <StyledComponent isRed />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches new styles when a new style collector is supplied', () => {
        interface ToggleProps extends Props {
            isAlternate?: boolean;
        }

        const styles = styleCollector<ToggleProps>('foo').element`
            background-color: blue;
        `.modifier(true)`
            background-color: red;
        `;

        const stylesAlt = styleCollector('foo').element`
            background-color: purple;
        `.modifier(true)`
            background-color: green;
        `;

        const StyledComponent: React.FC<ToggleProps> = props => {
            let activeStyles = styles;

            if (props.isAlternate) {
                activeStyles = stylesAlt;
            }

            const classNames = useStyles(activeStyles);

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent isRed />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();

        render(<StyledComponent isRed isAlternate />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches single style to the head', () => {
        const StyledComponent = () => {
            const classNames = useStyles(css`
                background-color: blue;
            `);

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches themed single style to the head', () => {
        interface Theme {
            backgroundColor: string;
        }

        const theme = {
            backgroundColor: 'Atomic Tangerine',
        };
        const StyledComponent = () => {
            const classNames = useStyles(
                css<Theme>`
                    background-color: ${theme => theme.backgroundColor};
                `,
            );

            return <span className={classNames}>Ahoy!</span>;
        };

        render(
            <ThemeProvider theme={theme}>
                <StyledComponent />
            </ThemeProvider>,
        );

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches object style to the head', () => {
        const StyledComponent = () => {
            const classNames = useStyles({
                backgroundColor: 'blue',
            });

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches object style with pseudo states to the head', () => {
        const StyledComponent = () => {
            const classNames = useStyles({
                backgroundColor: 'blue',
                ':hover': {
                    backgroundColor: 'brown',
                },
            });

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches object style with media queries to the head', () => {
        const StyledComponent = () => {
            const classNames = useStyles({
                backgroundColor: 'blue',
                '@media (min-width: 768px)': {
                    backgroundColor: '#deecff',
                },
            });

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('attaches object style with keyframe animations to the head', () => {
        const StyledComponent = () => {
            const classNames = useStyles({
                //@ts-ignore
                backgroundColor: 'blue',
                '@keyframes rotating': {
                    //@ts-ignore
                    from: {
                        transform: 'rotate(0deg)',
                    },
                    to: {
                        transform: 'rotate(360deg)',
                    },
                },
            });

            return <span className={classNames}>Ahoy!</span>;
        };

        render(<StyledComponent />);

        expect(document.getElementsByTagName('head')[0]).toMatchSnapshot();
    });

    it('object styles with different interpolations should be mounted separately', () => {
        const Box: FC<{ testId: string; size?: number }> = ({
            testId,
            size = 40,
        }) => {
            const classNames = useStyles({
                height: `${size}px`,
                width: `${size}px`,
            });

            return <div data-testid={testId} className={classNames} />;
        };

        const { getByTestId } = render(
            <Fragment>
                <Box testId="box-1" />
                <Box testId="box-2" size={20} />
            </Fragment>,
        );

        expect(getByTestId('box-1')!.className).toEqual('css__24024913390');
        expect(getByTestId('box-2')!.className).not.toEqual('css__24024913390');
    });

    it('css styles with different interpolations should be mounted separately', () => {
        const Box: FC<{ testId: string; size?: number }> = ({
            testId,
            size = 40,
        }) => {
            const classNames = useStyles(css`
                height: ${size}px;
                width: ${size}px;
            `);

            return <div data-testid={testId} className={classNames} />;
        };

        const { getByTestId } = render(
            <Fragment>
                <Box testId="box-1" />
                <Box testId="box-2" size={20} />
            </Fragment>,
        );

        expect(getByTestId('box-1')!.className).toEqual('css__24402206290');
        expect(getByTestId('box-2')!.className).not.toEqual('css__24402206290');
    });
});
