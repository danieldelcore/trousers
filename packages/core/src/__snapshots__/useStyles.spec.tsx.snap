// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyles attaches new styles when a new style collector is supplied 1`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--491189580"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--491189580{background-color:red;}

  </style>
</head>
`;

exports[`useStyles attaches new styles when a new style collector is supplied 2`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--491189580 .foo__3016956831 .foo--3826246206"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--491189580{background-color:red;}

    .foo__3016956831{background-color:purple;}

    .foo--3826246206{background-color:green;}

  </style>
</head>
`;

exports[`useStyles attaches single style to the head 1`] = `
<head>
  <style
    data-trousers=".css__293439549"
    type="text/css"
  >
    .css__293439549{background-color:blue;}

  </style>
</head>
`;

exports[`useStyles attaches styles to the head 1`] = `
<head>
  <style
    data-trousers=".foo__3889083325"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

  </style>
</head>
`;

exports[`useStyles attaches styles with modifiers to the head 1`] = `
<head>
  <style
    data-trousers=".foo__3889083325"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

  </style>
</head>
`;

exports[`useStyles attaches styles with modifiers to the head 2`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--491189580"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--491189580{background-color:red;}

  </style>
</head>
`;

exports[`useStyles attaches styles with multiple named modifiers to the head 1`] = `
<head>
  <style
    data-trousers=".foo__3889083325"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

  </style>
</head>
`;

exports[`useStyles attaches styles with multiple named modifiers to the head 2`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--primary491189580"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--primary491189580{background-color:red;}

  </style>
</head>
`;

exports[`useStyles attaches styles with multiple named modifiers to the head 3`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--primary491189580 .foo--secondary112991879"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--primary491189580{background-color:red;}

    .foo--secondary112991879{background-color:isBlue;}

  </style>
</head>
`;

exports[`useStyles attaches styles with multiple named modifiers to the head 4`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--primary491189580 .foo--secondary112991879"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--primary491189580{background-color:red;}

    .foo--secondary112991879{background-color:isBlue;}

  </style>
</head>
`;

exports[`useStyles attaches styles with named modifiers to the head 1`] = `
<head>
  <style
    data-trousers=".foo__3889083325"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

  </style>
</head>
`;

exports[`useStyles attaches styles with named modifiers to the head 2`] = `
<head>
  <style
    data-trousers=".foo__3889083325 .foo--primary491189580"
    type="text/css"
  >
    .foo__3889083325{background-color:blue;}

    .foo--primary491189580{background-color:red;}

  </style>
</head>
`;

exports[`useStyles attaches themed single style to the head 1`] = `
<head>
  <style
    data-trousers=".css__42468333793516409327"
    type="text/css"
  >
    .css__42468333793516409327{background-color:Atomic Tangerine;}

  </style>
</head>
`;

exports[`useStyles attaches themed styles to the head 1`] = `
<head>
  <style
    data-trousers=".foo__38783974113106931827"
    type="text/css"
  >
    .foo__38783974113106931827{background-color:purple;}

  </style>
</head>
`;

exports[`useStyles attaches themed styles to the head 2`] = `
<head>
  <style
    data-trousers=".foo__38783974113106931827 .foo--38783974113106931827"
    type="text/css"
  >
    .foo__38783974113106931827{background-color:purple;}

    .foo--38783974113106931827{background-color:purple;}

  </style>
</head>
`;
